src/core/reaction_ensemble.cpp

```diff
@@ -1497,6 +1497,18 @@ int ConstantpHEnsemble::get_random_valid_p_id() {
  */
 int ConstantpHEnsemble::do_reaction(int reaction_steps) {
 
+  bool reaction_possible = false;
+
+  for (int reaction_id = 0; reaction_id < reactions.size(); reaction_id++) {
+    if (all_reactant_particles_exist(reaction_id)) {
+      reaction_possible = true;
+    }
+  }
+
+  if (reaction_possible == false) {
+    return 1;
+  }
+
```

Comment:
isn't it possible to get this kind of logic implicitly? If no particle of the reactant type if found nothing should happen automatically without manually checking?

Reply:
The problem that causes the infinite loop has its source in the while loop in lines 1521-1546. This while loop searches  the particle ids that have a type that matches the reactant/product types. Of course, if you do not create such particle, then this loop runs for ever.
As you suggest, another possible solution would be to do this task without using a while statement so that if there is no reactant then the program does nothing. However, in my opinion, manually checking it has the advantage that we can integrate an exception error preventing the user from doing such simulation (since I cannot think where one would attempt to run the algorithm without reactants if it is not by mistake).

Reply:
The problem that causes the infinite loop has its source in the while loop in lines 1521-1546. This while loop searches the particle ids that have a type that matches the reactant/product types. Of course, if you do not create such particle, then this loop runs for ever.
As you suggest, another possible solution would be to do this task without using a while statement so that if there is no reactant then the program does nothing. However, in my opinion, manually checking it has the advantage that we can integrate an exception error preventing the user from doing such simulation (since I cannot think where one would attempt to run the algorithm without reactants if it is not by mistake).

```

Reply:
The point is that such situation should not occur. If it occurs, then it is an error in the system setup, so it cannot be a simple no-op. It should raise an error.

Reply:
in my mind the reaction algorithm itself should be independent of the state of the system. I can also define a non bonded interaction of two non existing particle types without having an infinite loop. The reaction of reactants is also some kind of particle interaction, right? Why should it be special here?

Reply:
All methods default to a no-op when not enough particles are present, except for `WidomInsertion` which throws an error. I don't immediately see why it should be an issue to not have enough particles of type HA if there is at least one particle of type A-. Could you please clarify this for us?
Also if you reverse the reactant types with the adduct types, you will get a bug in the other direction, if I understand e4104d28e6 correctly.

Reply:
@jngrad  Considering the reaction HA <--> A- + H+; If you don't have any particle HA, in addition of having at least one particle of type A-  you also need at least a counter-ion particle H+ so the reaction is possible in the reverse way. In the current develop version creating either only A- or H+ particles leads to an infinite loop error because the program searches in vain particles matching the missing type. You are right, in e4104d2  I introduced the reverse bug, I think that I have solved the problem in the most recent commit.
@KaiSzuttor If I understood you right, you suggest that it could be defined that a reaction is possible between some particle types even if they are not actually created in the system. In this case however, what the user is attempting is to actually perform the reaction, which is impossible if there are not enough particle reactants or products.

Reply:
To elaborate on @KaiSzuttor's remark, shouldn't it be possible to use the Constant pH method with competitive reactions? For example HA <--> A- + H+ together with A- + BX <--> AB + X- so that reaction B consumes all AH (by driving the acid-base reaction forward) and leaves HX as the acid. One application would be the S<sub>N</sub>2 reaction between HBr and alkyl chlorides.

Reply:
@jngrad  thank you for your clarification, now I understand your point. Now I agree with you that the desired output would be a no-op so one could couple the acid/base reaction with other reactions

Reply:
Indeed, @jngrad is right that all reactants could be consumed if different reactions are competing for the same product. However, I would discourage anybody from intentionally using RxMC in this regime. It suffers from serious finite-size artifacts if the number of (any type of) reacting particles is on the order of unity. It is related to the fact that particle numbers in a simulation have to be integers whereas most existing theories work with concentrations as continuous variables.

Reply:
Note: SN2 refers to a specific reaction mechanism. A reaction of the type A- + BX <--> AB + X- is not compatible with the way currently reactions are implemented in Espresso. The current implementation allows to create or destroy particles at random positions but it does not allow to transfer a particle between two binding partners. It is also one of the points on our wishlist.

Reply:
Be aware it definitely depends on what you mean:
Of course it is totally valid to create this reaction 
A+B <-> C+D 
Whit artbitrary A, B, C, D. I.e. also :
 A- + BX <--> AB + X-
However all the types will be "primitive" beads i.e. not have bonds in the bead "BX". But it is completely fine for the "BX"  type to have different lenard jones parameters (i.e. appear bigger) etc...
Maybe this would require having a type dependent exclusion radius for some more extreme variations...

Reply:
> A reaction of the type A- + BX <--> AB + X- is not compatible with the way currently reactions are implemented in Espresso. The current implementation allows to create or destroy particles at random positions but it does not allow to transfer a particle between two binding partners.

Good point. I also thought about the case of the titration curve of a buffer solution, i.e. a mixture of a weak acid and a strong acid. Titration by a strong base can trigger the bug once the buffer capacity is used up. But that's an edge case, one probably wouldn't want to add more base than there is buffer.